### ISSUE TITLE
provided_by, is_defined_by changes

### ISSUE THREAD

**Issue by edeutsch** [RTXteam/RTX#1274]
There are changes afoot with provided_by, is_defined_by within the TRAPI and EPc working groups. The final form is not settled yet. but the feet are shuffling toward knowledge_provider, original_source.

@amykglen when you are in this part of the code next, I suggest we change:
```
          "attributes": [
            {
              "name": "provided_by",
              "value": [
                "SEMMEDDB:"
              ],
              "type": "biolink:provided_by",
              "url": null,
              "source": null
            },
             {
              "name": "is_defined_by",
              "value": "ARAX/KG2c",
              "type": "biolink:Unknown",
              "url": null,
              "source": null
            }
          ]
```

to:

```
          "attributes": [
            {
              "name": "original_source",                <--- likely new name
              "value": [
                "SEMMEDDB"                      <--- I removed colon
              ],
              "type": "biolink:original_source",               <--- likely new name
              "url": null,                                     <--- Do we have a URL for SEMMEDB?
              "source": null
            },
            {
              "name": "knowledge_provider",               <--- likely new name
              "value": "RTX KG2",                          <--- What is our KP branding string? See #1129 
              "type": "biolink:knowledge_provider",               <--- likely new name
              "url": null,                                               <--- Do we have a URL for RTX KG2?
              "source": null
            }
          ]
```

Do we like these changes? there is still an opportunity to shape the new new design

**Comment by amykglen** [comment]
nice, I like these changes - makes it much more clear what the attributes are intended to capture.

I started switching `expand` to this in a branch (`issue1274`). I think `overlay` will also need to be updated, as it also adds edges to the KG with `provided_by`/`is_defined_by` attributes. (@finnagin)

re: the colon you removed from the provided_by value of "SEMMEDDB:" - I think @saramsey / @ericawood / @kvarforl intentionally put that colon there in KG2, so I'd suggest we leave it in or discuss with them whether it should be there.

**Comment by ericawood** [comment]
Regarding URLs for the sources, kg2's `curies-to-urls-map.yaml` (https://github.com/RTXteam/RTX/blob/master/code/kg2/curies-to-urls-map.yaml) connects the sources with their urls.

**Comment by amykglen** [comment]
oh nice, thanks, @ericawood! so maybe `expand` could use that file to get a url for each provided_by curie.

although I'm not sure how to stuff that into an `Attribute` since `provided_by` is a list, and each item in that list would have a different URL, but an `Attribute` only has one URL slot...

**Comment by edeutsch** [comment]
Three things:
1) The 'url' is *not* intended as an expansion of the CURIE. That would be redundant, since all CURIEs are just compact URIs. The 'url' is intended as a human-consumable how do I learn more about this attribute value . In this case, it would be a human-consumable URL to learn more about SEMMEDDB or RTX KG2. i.e. a URL describing the value.
2) There was some talk about making 'url' a list, but that is unsettled
3) In fact, this whole schema is somewhat in flux. It will be different for TRAPI 1.1, but probably only slightly. Battle will ensue at next week's TRAPI call.

**Comment by amykglen** [comment]
re: 1. hm - so "SEMMEDDB:" expands to "https://skr3.nlm.nih.gov/SemMedDB" in the yaml file Erica shared. isn't that where one would want to go to learn more about SEMMEDDB? that's what I was getting at.

**Comment by edeutsch** [comment]
ah, I see. Well, I'm thinking that that may work for a few. But I think most of those are intended as expansions for CURIE prefixes and so for example for Drugbank, it's
https://identifiers.org/drugbank:
and sending a person there isn't so useful

**Comment by saramsey** [comment]
> Three things:
> 
>     1. The 'url' is _not_ intended as an expansion of the CURIE. That would be redundant, since all CURIEs are just compact URIs. The 'url' is intended as a human-consumable how do I learn more about this attribute value . In this case, it would be a human-consumable URL to learn more about SEMMEDDB or RTX KG2. i.e. a URL describing the value.
> 
>     2. There was some talk about making 'url' a list, but that is unsettled
> 
>     3. In fact, this whole schema is somewhat in flux. It will be different for TRAPI 1.1, but probably only slightly. Battle will ensue at next week's TRAPI call.

Eric, thank you for bringing this to my attention. So, the Biolink model has a quite different take on the semantics of the `iri` node slot. Quoting directly from the Biolink yaml file:

<img width="721" alt="Screen Shot 2021-04-13 at 1 18 03 PM" src="https://user-images.githubusercontent.com/5562107/114615670-cfaad080-9c5a-11eb-9f2d-0f5669dd1d9f.png">

https://github.com/biolink/biolink-model/blob/8b5ca35ad171f5837ae3b61d3df31e5c3a04b344/biolink-model.yaml#L465

It has been so since before Biolink was versioned, in fact since the *initial commit* 3 years ago. I want to stress the following: in designing the KG2 build system, a great deal of design effort and testing/validation effort was invested in ensuring consistency between a node's `iri` and its `id` CURIE, following the Biolink metamodel standard. Now admittedly, I don't have a strong direct stake in the content or semantics of the `url` attribute in TRAPI, but I have a major stake in, and opinion about, the `iri` node slot in the Biolink metamodel. If ARAX can set the `uri` attribute for a given node (`node_n`) in the return graph by doing a one-hop lookup in which ARAX takes the `provided_by` property value (which is a CURIE) from `node_n` and then grabs the `iri` property of the node (we'll call it `ont-node`) whose `id` matches the `provided_by` property of `node_n`), I'm fine with it.

I really really hope we are not proposing to change the semantics of the `iri` slot of a node in the Biolink metamodel.

**Comment by saramsey** [comment]
I think in KG2, there should be no situation where a node `iri` property or an expanded `provided_by` CURIE results in the URI
```
https://identifiers.org/drugbank:
```
The `provided_by` CURIE for the DrugBank knowledge source (i.e., representing all of DrugBank) should be
```
identifiers_org_registry:drugbank
```
which should expand to:
```
https://registry.identifiers.org/registry/drugbank
```
You can see for yourself in KG2.5.2,
<img width="845" alt="Screen Shot 2021-04-13 at 3 43 44 PM" src="https://user-images.githubusercontent.com/5562107/114630258-0db1ef80-9c6f-11eb-8061-308bb8abd5cc.png">

On the other hand, an *individual* DrugBank drug concept, like `DRUGBANK:DB00316` should get expanded to:
```
https://identifiers.org/drugbank:DB00316
```
which resolves dynamically to
```
https://go.drugbank.com/drugs/DB00316
```
as shown here:
<img width="807" alt="Screen Shot 2021-04-13 at 3 44 58 PM" src="https://user-images.githubusercontent.com/5562107/114630335-3803ad00-9c6f-11eb-9166-168a6546ae79.png">

I'm not seeing a problem here, with how DrugBank IRIs are represented in KG2. Can someone please help me understand?  Thanks.

**Comment by edeutsch** [comment]
Hi Steve, I suppose I'm confused by your posts here.
The 'url' property in Attribute is not intended to be the concept's 'iri'?
Is that the question?

**Comment by edeutsch** [comment]
Here's my best example of an Attribute:
```
    {
      "attribute_type_id": "biolink:knowledge_provider_source",  
      "original_attribute_name": "knowledge provider source",
      "attribute_source": "infores:rtxkg2_kp",
      "value": "infores:rtxkg2_kp",       
      "value_type_id": "biolink:InformationResource",     
      "value_url":  "https://github.com/RTXteam/RTX/tree/master/code/kg2",
      "description": "NCATS Translator Knowledge Provider RTX KG2",
    },
```

This is not yet settled. Sparks will fly this Thursday at 10am.

**Comment by saramsey** [comment]
Thanks, @edeutsch. I suppose I was just concerned that there might be a movement to change the `iri` property (which after all, is just the i18n-friendly version of a "uri") in Biolink. Never mind, then. ;-)

**Comment by saramsey** [comment]
And in that case, if the Attribute `uri` has nothing to do with any field in KG2, I defer to the folks working on the more TRAPI-proximal parts of the system, as to what should be the team's position on this proposed TRAPI change.

**Comment by edeutsch** [comment]
1. It's not a `uri`, it's a `url`. It is not intended to be an identifier. it is intended to be a plain-old web page.
2. I think it is fair to conclude that no field in KG2 should be used to populate the `url` property. It is something that ARAX code should populate basic on its logic.
3. I don't understand which proposed TRAPI change you mean

**Comment by saramsey** [comment]
>     1. It's not a `uri`, it's a `url`. It is not intended to be an identifier. it is intended to be a plain-old web page.

A URL *has* to be a URI, by definition, per [RFC 3986](https://tools.ietf.org/html/rfc3986).

**Comment by saramsey** [comment]
> I don't understand which proposed TRAPI change you mean

I was referring to this change

> The final form is not settled yet. but the feet are shuffling toward `knowledge_provider`, `original_source`.

**Comment by amykglen** [comment]
can this issue be closed since #1414 now captures these changes?
